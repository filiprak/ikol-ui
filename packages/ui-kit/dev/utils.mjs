import { existsSync } from 'fs';
import { readFile, writeFile, mkdir } from 'fs/promises';
import path from 'path';
import { fileURLToPath } from 'url';

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

export async function useTemplateFile(filename, outputFilepath, variables) {
  const data = await readFile(path.resolve(__dirname, 'templates', filename), 'utf8');
  return useTemplate(data, outputFilepath, variables);
}

export async function useTemplate(templateStr, outputFilepath, variables) {
  const updatedTemplate = templateStr.replace(/\[\[([a-zA-Z0-9_]+)\]\]/g, (match, key) => {
    return variables[key] || match;
  });

  if (!existsSync(path.dirname(outputFilepath))) {
    await mkdir(path.dirname(outputFilepath));
  }
  await writeFile(outputFilepath, updatedTemplate);
}

export const kebabize = str => {
  return str.split('').map((letter, idx) => {
      return letter.toUpperCase() === letter
          ? `${idx !== 0 ? '-' : ''}${letter.toLowerCase()}`
          : letter;
  }).join('');
}

export const ROOT_DIR = path.resolve(__dirname, '..');
