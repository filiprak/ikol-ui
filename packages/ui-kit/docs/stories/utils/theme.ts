import { useTheme } from '@ui/composables';
import { createIkolUI } from '@ui/index';
import { App } from 'vue';
import { StoryContext, VueRenderer } from '@storybook/vue3';

export function setupTheme(context: any) {
    const theme = useTheme();

    theme.is_dark.value = context?.globals?.theme === 'dark';

    return theme;
}

export function setupApp(app: App, context?: StoryContext<VueRenderer>) {
    app.use(createIkolUI());
}
