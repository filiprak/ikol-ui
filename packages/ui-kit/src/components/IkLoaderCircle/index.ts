import IkLoaderCircle from './IkLoaderCircle.vue';
export interface IkLoaderCircle extends InstanceType<typeof IkLoaderCircle> { };
export { IkLoaderCircle };
