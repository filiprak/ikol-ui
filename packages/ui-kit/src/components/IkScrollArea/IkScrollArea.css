/* stylelint-disable selector-class-pattern */

.ik-scroll-area {
    width: 100%;
    min-height: 0;
    max-height: 100%;
}

.ik-scroll-area--no-auto-anchor > div > div > div > .simplebar-content-wrapper {
    /* fixes SimpleBar invalid scroll behaviour bug #3961 */
    overflow-anchor: none !important;
}

.ik-scroll-area--native {
    position: relative;
    overflow: auto;
}

.ik-scroll-area--hide-x > .simplebar-track.simplebar-horizontal,
.ik-scroll-area--hide-y > .simplebar-track.simplebar-vertical {
    display: none;
}

.ik-scroll-area--hide-y {
    overflow: auto hidden;
}

.ik-scroll-area__loader {
    margin: 10px;
    text-align: center;
}

.ik-scroll-area > .simplebar-track .simplebar-scrollbar.simplebar-visible:before {
    opacity: 0.25;
    background-color: var(--scrollbar-color);
}

.ik-scroll-area > .simplebar-track.simplebar-vertical.simplebar-hover {
    width: 17px;
    background-color: var(--scrollbar-track-color);
}

.ik-scroll-area > .simplebar-track.simplebar-horizontal.simplebar-hover {
    height: 17px;
    background-color: var(--scrollbar-track-color);
}

.ik-scroll-area > .simplebar-track.simplebar-horizontal.simplebar-hover .simplebar-scrollbar:before {
    height: 13px;
}

.ik-scroll-area__shadow-right,
.ik-scroll-area__shadow-left {
    position: absolute;
    height: 100%;
    width: 100px;
    display: flex;
    align-items: center;
    z-index: 1;
    pointer-events: none;
    padding: 0 10px;
}

.ik-scroll-area__shadow-right .ik-button,
.ik-scroll-area__shadow-left .ik-button {
    pointer-events: all;
    box-shadow: 1px 2px 3px 0 var(--shadow-color-1);
}

.ik-scroll-area__shadow-right {
    -webkit-box-shadow: inset -45px 0 25px -10px rgba(255, 255, 255, 0.8);
    -moz-box-shadow: inset -45px 0 25px -10px rgba(255, 255, 255, 0.8);
    box-shadow: inset -45px 0 25px -10px rgba(255, 255, 255, 0.8);
    justify-content: flex-end;
    right: 0;
}


.ik-scroll-area__shadow-left {
    -webkit-box-shadow: inset 45px 0 25px -10px rgba(255, 255, 255, 0.8);
    ;
    -moz-box-shadow: inset 45px 0 25px -10px rgba(255, 255, 255, 0.8);
    ;
    box-shadow: inset 45px 0 25px -10px rgba(255, 255, 255, 0.8);
    justify-content: flex-start;
    left: 0;
}

.ik-scroll-area .fade-enter-active,
.ik-scroll-area .fade-leave-active {
    transition: opacity .5s;
}

.ik-scroll-area .fade-enter,
.ik-scroll-area .fade-leave-to {
    opacity: 0;
}