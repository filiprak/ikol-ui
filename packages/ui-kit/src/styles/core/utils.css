/* shadow */
.ik-shadow-0 {
    box-shadow: none;
}

.ik-shadow-1 {
    box-shadow: 0 1px 3px 0 var(--shadow-color-1);
}

.ik-shadow-2 {
    box-shadow: 0 2px 5px 0 var(--shadow-color-1);
}

.ik-shadow-3 {
    box-shadow: 0 2px 10px 0 var(--shadow-color-1);
}

.ik-shadow-4 {
    box-shadow: 0 2px 15px 0 var(--shadow-color-1);
}

.ik-shadow-5 {
    box-shadow: 0 2px 20px 0 var(--shadow-color-1);
}

.ik-shadow-6 {
    box-shadow: 0 3px 30px 0 var(--shadow-color-1);
}

/* skeleton background */
@-webkit-keyframes skeleton-loading {
    to {
        transform: translateX(100%)
    }
}

@keyframes skeleton-loading {
    to {
        transform: translateX(100%)
    }
}

.ik-bg-skeleton {
    background: var(--skeleton-color-1);
    overflow: hidden;
    position: relative;
    user-select: none;
}

.ik-bg-skeleton:after {
    background: linear-gradient(90deg, hsla(0, 0%, 100%, 0), var(--skeleton-color-2), hsla(0, 0%, 100%, 0));
    -webkit-animation: skeleton-loading 1.5s infinite;
    animation: skeleton-loading 1.5s infinite;
    content: '';
    height: 100%;
    left: 0;
    position: absolute;
    right: 0;
    top: 0;
    transform: translateX(-100%);
    z-index: 1
}

.ik-cursor--pointer {
    cursor: pointer;
}

/* design utils */
.ik-design--bg-primary {
    background-color: var(--primary-color);
    color: var(--on-primary-color);
}

.ik-design--bg-success {
    background-color: var(--success-color);
    color: var(--on-success-color);
}

.ik-design--bg-error {
    background-color: var(--error-color);
    color: var(--on-error-color);
}

.ik-design--bg-accent {
    background-color: var(--accent-color);
    color: var(--on-accent-color);
}

.ik-design--bg-secondary {
    background-color: var(--secondary-color);
    color: var(--on-secondary-color);
}

.ik-design--bg-default {
    background-color: var(--surface-darken-color);
    color: var(--on-surface-darken-color-3);
}

/* height */
.ik-height--full {
    height: 100%;
}

.ik-width--full {
    width: 100%;
}
