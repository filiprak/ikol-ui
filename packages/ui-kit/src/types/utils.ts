import type { ExtractPropTypes } from 'vue';

/* eslint-disable @typescript-eslint/naming-convention */
export type IkExtractProps<T> =
    T extends new (...angs: any) => { $props: infer P; } ? NonNullable<P> : // eslint-disable-line @typescript-eslint/no-explicit-any
    T extends (props: infer P, ...args: any) => any ? P : // eslint-disable-line @typescript-eslint/no-explicit-any
    {};

export type IkExtractPropTypes<T> = Prettify<Readonly<ExtractPropTypes<T>>>;

export type IkExtractExposed<T> =
    T extends new (...angs: any) => infer E ? E : // eslint-disable-line @typescript-eslint/no-explicit-any
    T extends (props: any, ctx: any, expose: (exposed: infer E) => any, ...args: any) => any ? NonNullable<E> : // eslint-disable-line @typescript-eslint/no-explicit-any
    {};

export type IkKeysByType<T, U> = keyof {
    [K in keyof T as T[K] extends U ? K : never]: T[K];
};

export enum IkUIDesignColor {
    primary = 'primary',
    success = 'success',
    error = 'error',
    accent = 'accent',
    default = 'default',
}

export enum IkUIElemSize {
    xs = 'xs',
    sm = 'sm',
    md = 'md',
    lg = 'lg',
}

export enum IkUIVariant {
    flat = 'flat',
    filled = 'filled',
    outline = 'outline',
    pill = 'pill',
}

export enum IkPosition {
    top = 'top',
    right = 'right',
    bottom = 'bottom',
    left = 'left',
    horizontal = 'horizontal',
    vertical = 'vertical',
}

export enum IkAlignment {
    start = 'start',
    center = 'center',
    end = 'end',
}

export type EnumT = Record<string, string> | Record<string | number, string | number>;
export type ObjValueT<T> = T[keyof T];
export type DefinedKeys<T> = { [K in keyof T]: undefined extends T[K] ? never : K }[keyof T];
export type OmitUndefined<T> = { [K in DefinedKeys<T>]: T[K] };
export type NotArray<T> = T extends any[] ? never : T; // eslint-disable-line @typescript-eslint/no-explicit-any
export type DefinedComponent = new (...args: any[]) => any; // eslint-disable-line @typescript-eslint/no-explicit-any
export type Prettify<T> = {
    [K in keyof T]: T[K];
} & {};
