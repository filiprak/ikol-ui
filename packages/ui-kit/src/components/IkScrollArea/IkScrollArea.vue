<template>
    <div ref="root_el"
         :class="root_class"
         data-simplebar
         data-simplebar-auto-hide="false">
        <div class="simplebar-wrapper">
            <template v-if="horizontal_arrows">
                <transition name="fade">
                    <div v-if="arrows.left"
                         class="ik-scroll-area__shadow-left">
                        <IkButton circle
                                  filled
                                  icon="arrow-left"
                                  size="lg"
                                  @click="_onArrowClick(-1)" />
                    </div>
                </transition>
                <transition name="fade">
                    <div v-if="arrows.right"
                         class="ik-scroll-area__shadow-right">
                        <IkButton circle
                                  filled
                                  icon="arrow-right"
                                  size="lg"
                                  @click="_onArrowClick(1)" />
                    </div>
                </transition>
            </template>
            <div class="simplebar-height-auto-observer-wrapper">
                <div class="simplebar-height-auto-observer" />
            </div>
            <div class="simplebar-mask">
                <div class="simplebar-offset">
                    <div ref="scroll_el"
                         class="simplebar-content-wrapper"
                         @scrollend="_onScrollEnd"
                         @scroll.passive="_onScroll">
                        <div ref="content_el"
                             class="simplebar-content">
                            <slot v-if="!reverse"></slot>
                            <div v-if="show_loader"
                                 class="ik-scroll-area__loader">
                                <IkLoaderCircle indeterminate
                                                :design="loader_design || 'default'"
                                                :size="20"
                                                :thickness="2" />
                            </div>
                            <slot v-if="reverse"></slot>
                        </div>
                    </div>
                </div>
            </div>
            <div class="simplebar-placeholder" />
        </div>
        <div class="simplebar-track simplebar-horizontal">
            <div class="simplebar-scrollbar" />
        </div>
        <div class="simplebar-track simplebar-vertical">
            <div class="simplebar-scrollbar" />
        </div>
    </div>
</template>
<script setup lang="ts">
import '@ui/styles';
import 'simplebar/dist/simplebar.min.css';
import './IkScrollArea.css';
import SimpleBar from 'simplebar';
import { computed, nextTick, onBeforeUnmount, onMounted, reactive, ref, useAttrs } from 'vue';
import { IkLoaderCircle } from '@ui/components/IkLoaderCircle';
import { IkButton } from '@ui/components/IkButton';
import { isFunction } from '@ui/utils/helpers';

const SCROLL_EL_DATA_KEY = 'ik-scroll-area';

export interface HTMLIkScrollElement extends HTMLElement {
    [SCROLL_EL_DATA_KEY]?: {
        is_bottom?: boolean,
        last_offset?: number | null,
        last_scroll_top?: number,
        last_scroll_height?: number,
        last_bottom_distance?: number,
    }
}
const emit = defineEmits({});
const attrs = useAttrs();
const props = withDefaults(
    defineProps<{
        hide_x?: boolean,
        hide_y?: boolean,
        disable_auto_anchoring?: boolean,
        loading?: boolean,
        hide_loader?: boolean,
        loader_design?: string,
        bound_threshold?: number,
        reverse?: boolean,
        pagination?: { offset: number, has_more: boolean },
        horizontal_arrows?: boolean,
        horizontal_arrow_step?: number,
    }>(),
    {
        hide_x: false,
        hide_y: false,
        disable_auto_anchoring: false,
        loading: false,
        hide_loader: false,
        bound_threshold: 20,
        reverse: false,
        pagination: () => ({
            offset: 0,
            has_more: false,
        }),
        horizontal_arrows: false,
    },
);

const root_el = ref<HTMLElement>();
const scroll_el = ref<HTMLElement>();
const content_el = ref<HTMLElement>();

const arrows = reactive({
    left: false,
    right: false,
});

const root_class = computed(() => {
    return {
        'ik-scroll-area': true,
        'ik-scroll-area--hide-x': props.hide_x,
        'ik-scroll-area--hide-y': props.hide_y,
        'ik-scroll-area--no-auto-anchor': props.disable_auto_anchoring,
    };
});
const infinite = computed(() => {
    return !!attrs.onIkLoadMore;
});
const show_loader = computed(() => {
    return !props.hide_loader && infinite.value && props.pagination.has_more;
});

function getContentHeight() {
    const el = content_el.value;
    if (el) {
        const rect = el.getBoundingClientRect();
        return rect.height;
    } else {
        return null;
    }
}
function scrollTop(offset: number, animate = false) {
    const el = getScrollEl();
    if (el) {
        el.scroll({
            top: (offset || 0),
            behavior: animate ? 'smooth' : undefined,
        });
    }
}
function scrollLeft(offset: number, animate = false) {
    const el = getScrollEl();
    if (el) {
        el.scroll({
            left: (offset || 0),
            behavior: animate ? 'smooth' : undefined,
        });
    }
}
function getScrollWidth() {
    const el = getScrollEl();
    return el ? el.scrollWidth : null;
}
function getScrollTop() {
    const el = getScrollEl();
    return el ? el.scrollTop : null;
}
function getScrollLeft() {
    const el = getScrollEl();
    return el ? el.scrollLeft : null;
}
function getScrollTopMax() {
    const el = getScrollEl();
    return el ? el.scrollHeight - el.clientHeight : null;
}
function getScrollLeftMax() {
    const el = getScrollEl();
    return el ? el.scrollWidth - el.clientWidth : null;
}
function getScrollLeftRemaining() {
    const left_max = getScrollLeftMax();
    const left = getScrollLeft();

    if (left_max !== null && left !== null) {
        return left_max - left;
    } else {
        return null;
    }
}
function scrollBottom() {
    const el = getScrollEl();
    if (el) {
        nextTick(() => {
            el.scrollTop = el.scrollHeight;
        });
    }
}
function scrollTo(el: HTMLElement, align: 'center' | 'top') {
    const scroll_el = getScrollEl();

    if (scroll_el && el) {
        nextTick(() => {
            const s_rect = scroll_el.getBoundingClientRect();
            const rect = el.getBoundingClientRect();
            let top = 0;
            switch (align) {
                case 'center':
                    top = Math.floor(el.offsetTop + (rect.height / 2) - (s_rect.height / 2));
                    break;
                default:
                    top = 0;
            }
            scroll_el.scrollTop = Math.max(top, 0);
        });
    }
}
function getScrollEl(): HTMLIkScrollElement | undefined {
    return scroll_el.value;
}
function _getContentEl() {
    return content_el.value;
}
function _onArrowClick(steps: number) {
    scrollLeft((getScrollLeft() || 0) + (steps * (props.horizontal_arrow_step || 300)), true);
}
function _onScroll(event: Event) {
    const el = event.target as HTMLIkScrollElement | null;

    if (infinite.value && el) {
        const data = el[SCROLL_EL_DATA_KEY] || {};
        const last_offset = data.last_offset;
        const from_bottom = el.scrollHeight - el.scrollTop - el.clientHeight;
        const is_bottom = from_bottom <= 10;

        if (!props.loading) {
            const reached_bound = props.reverse
                ? (el.scrollTop <= props.bound_threshold)
                : (el.scrollTop + el.clientHeight + props.bound_threshold >= el.scrollHeight);
            if (reached_bound) {
                const has_more_items = props.pagination.has_more;
                if (has_more_items && last_offset !== props.pagination.offset) {
                    data.last_offset = props.pagination.offset;
                    emit('ik-load-more', { pagination: props.pagination });
                }
            }
        }
        el[SCROLL_EL_DATA_KEY] = {
            is_bottom,
            last_offset: data.last_offset || null,
            last_scroll_height: el.scrollHeight,
            last_bottom_distance: el.scrollHeight - el.scrollTop - el.clientHeight,
        };
    }
    if (props.horizontal_arrows) {
        _updateArrows();
    }
    if (isFunction(attrs.onScroll)) {
        attrs.onScroll(event);
    }
}
function _onScrollEnd(event: Event) {
    if (isFunction(attrs.onScrollend)) {
        attrs.onScrollend(event);
    }
}
function _onContentChange() {
    const el = getScrollEl();

    if (props.reverse && el) {
        const data = el[SCROLL_EL_DATA_KEY] || {};
        const last_height = data.last_scroll_height || 0;

        nextTick(() => {
            if (data.is_bottom) {
                el.scrollTop = el.scrollHeight;
            } else {
                el.scrollTop = Math.max(el.scrollTop + (el.scrollHeight - last_height), 0);
            }
        });
    }
    if (props.horizontal_arrows) {
        _updateArrows();
    }
}
function _onScrollElReady(el: HTMLIkScrollElement) {
    if (props.reverse) {
        el.scrollTop = el.scrollHeight;
        el[SCROLL_EL_DATA_KEY] = {
            is_bottom: true,
            last_offset: null,
            last_scroll_top: el.scrollTop,
            last_scroll_height: el.scrollHeight,
            last_bottom_distance: el.scrollHeight - el.scrollTop - el.clientHeight,
        };
    }
    if (props.reverse || props.horizontal_arrows) {
        new MutationObserver(_onContentChange)
            .observe(el, {
                childList: true,
                subtree: true,
            });
        const r_obs = new ResizeObserver(_onContentChange);
        r_obs.observe(el);
        if (el !== _getContentEl()) {
            const c_el = _getContentEl();
            c_el && r_obs.observe(c_el);
        }
    }
}
function _updateArrows() {
    const left = getScrollLeft();
    const leftr = getScrollLeftRemaining();

    arrows.left = left !== null && (left > 2);
    arrows.right = leftr !== null && (leftr > 2);
}
onMounted(() => {
    const el = getScrollEl();

    if (el) {
        _onScrollElReady(el);
    }

    if (props.reverse) {
        nextTick(() => {
            scrollBottom();
        });
    }
    if (props.horizontal_arrows) {
        nextTick(() => {
            _updateArrows();
        });
    }
});

onBeforeUnmount(() => {
    const instance = SimpleBar.instances.get(root_el.value as Object);
    if (instance) {
        instance.unMount();
    }
});

defineExpose({
    getScrollEl,
    getContentHeight,
    scrollTop,
    scrollLeft,
    getScrollWidth,
    getScrollTop,
    getScrollTopMax,
    getScrollLeft,
    getScrollLeftMax,
    scrollBottom,
    scrollTo,
});
</script>
