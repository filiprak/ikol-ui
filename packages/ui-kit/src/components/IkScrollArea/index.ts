import type { IkExtractExposed } from '@ui/types/utils';
import IkScrollArea from './IkScrollArea.vue';
export interface IkScrollArea extends IkExtractExposed<typeof IkScrollArea> { };
export { IkScrollArea };
