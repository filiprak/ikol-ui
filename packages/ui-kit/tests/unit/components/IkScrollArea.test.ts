import type { RenderOptions } from '@ui/tests/tools/utils';
import { renderShallow } from '@ui/tests/tools/utils';
import { IkScrollArea } from '@ui/components/IkScrollArea';
import SimpleBar from 'simplebar';
import { IkLoaderCircle } from '@ui/components/IkLoaderCircle';

describe('IkScrollArea', () => {
    const getInstance = (options: RenderOptions<typeof IkScrollArea> = {}) => renderShallow(IkScrollArea, {}, options);

    it('renders with correct default state', () => {
        const wrapper = getInstance();
        expect(wrapper.classes()).toStrictEqual([
            'ik-scroll-area',
        ]);
        expect(wrapper.element.tagName).toBe('DIV');
        expect(wrapper.element).toMatchSnapshot();
    });

    it('renders with default slot', () => {
        const wrapper = getInstance({
            slots: {
                default: '<div class="test">Area scroll content...</div>',
            },
        });
        expect(wrapper.text()).toContain('Area scroll content...');
        expect(wrapper.element).toMatchSnapshot();
    });

    it('has simplebar attribute', () => {
        const wrapper = getInstance();
        expect(wrapper.attributes('data-simplebar')).toBe('');
    });

    it('calls SimpleBar.unMount() on destroy', () => {
        const wrapper = getInstance();

        const instance = new SimpleBar(wrapper.vm.$el);
        const unMount = jest.spyOn(instance, 'unMount');

        wrapper.unmount();

        expect(unMount).toHaveBeenCalled();
    });

    it('renders loader when loading and not all items loaded', () => {
        const wrapper = getInstance({
            props: {
                loading: true,
                pagination: {
                    offset: 8,
                    has_more: true,
                },
            },
            slots: {
                default: '<div class="test">Infinite scroll content...</div>',
            },
            attrs: {
                onIkLoadMore: jest.fn(),
            },
        });
        expect(wrapper.findComponent(IkLoaderCircle).isVisible()).toBe(true);
        expect(wrapper.text()).toContain('Infinite scroll content...');
        expect(wrapper.element).toMatchSnapshot();
    });

    it('does not render loader when all items loaded', () => {
        const wrapper = getInstance({
            props: {
                loading: true,
                pagination: {
                    offset: 9,
                    has_more: false,
                },
            },
            slots: {
                default: '<div class="test">Infinite scroll content...</div>',
            },
            attrs: {
                onIkLoadMore: jest.fn(),
            },
        });
        expect(wrapper.findComponent(IkLoaderCircle).exists()).toBe(false);
        expect(wrapper.text()).toContain('Infinite scroll content...');
    });

    it('emits event when scrolled bottom element', () => {
        const wrapper = getInstance({
            props: {
                loading: false,
                pagination: {
                    offset: 9,
                    has_more: true,
                },
            },
            slots: {
                default: '<div class="test">Infinite scroll content...</div>',
            },
            attrs: {
                onIkLoadMore: jest.fn(),
            },
        });

        wrapper
            .find('.simplebar-content-wrapper')
            .trigger('scroll');

        expect(wrapper.emitted('ik-load-more')).toStrictEqual([
            [expect.objectContaining({ pagination: expect.any(Object) })],
        ]);
    });

    it('does not emit event when loading', () => {
        const wrapper = getInstance({
            props: {
                loading: true,
                pagination: {
                    offset: 9,
                    has_more: true,
                },
            },
            slots: {
                default: '<div class="test">Infinite scroll content...</div>',
            },
            attrs: {
                onIkLoadMore: jest.fn(),
            },
        });

        wrapper
            .find('.simplebar-content-wrapper')
            .trigger('scroll');

        expect(wrapper.emitted('ik-load-more')).toBeUndefined();
    });

    it('does not emit event when all loaded', () => {
        const wrapper = getInstance({
            props: {
                loading: false,
                pagination: {
                    offset: 9,
                    has_more: false,
                },
            },
            slots: {
                default: '<div class="test">Infinite scroll content...</div>',
            },
            attrs: {
                onIkLoadMore: jest.fn(),
            },
        });

        wrapper
            .find('.simplebar-content-wrapper')
            .trigger('scroll');

        expect(wrapper.emitted('ik-load-more')).toBeUndefined();
    });

    it('renders with hide x and hide y', () => {
        const wrapper = getInstance({
            props: {
                hide_x: true,
                hide_y: true,
            },
        });

        expect(wrapper.classes()).toContain('ik-scroll-area--hide-x');
        expect(wrapper.classes()).toContain('ik-scroll-area--hide-y');
    });

    it('renders loading in reverse mode', () => {
        const wrapper = getInstance({
            props: {
                reverse: true,
                loading: true,
                pagination: {
                    offset: 9,
                    has_more: true,
                },
            },
            attrs: {
                onIkLoadMore: jest.fn(),
            },
            slots: {
                default: '<div class="test">Infinite scroll content...</div>',
            },
        });

        expect(wrapper.findComponent(IkLoaderCircle).isVisible()).toBeTruthy();
        expect(wrapper.element).toMatchSnapshot();
    });

    it('renders without loader', () => {
        const wrapper = getInstance({
            props: {
                hide_loader: true,
                loading: true,
                pagination: {
                    offset: 0,
                    has_more: true,
                },
            },
            attrs: {
                onIkLoadMore: jest.fn(),
            },
            slots: {
                default: '<div class="test">Infinite scroll content...</div>',
            },
        });

        expect(wrapper.findComponent(IkLoaderCircle).exists()).toBeFalsy();
    });

    it('renders with disabled native auto anchoring', () => {
        const wrapper = getInstance({
            props: {
                disable_auto_anchoring: true,
            },
        });

        expect(wrapper.classes()).toContain('ik-scroll-area--no-auto-anchor');
    });
});
