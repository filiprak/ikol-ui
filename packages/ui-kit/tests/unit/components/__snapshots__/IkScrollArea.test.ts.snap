// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`IkScrollArea renders loader when loading and not all items loaded 1`] = `
<div
  class="ik-scroll-area"
  data-simplebar=""
  data-simplebar-auto-hide="false"
>
  <div
    class="simplebar-wrapper"
  >
    <!--v-if-->
    <div
      class="simplebar-height-auto-observer-wrapper"
    >
      <div
        class="simplebar-height-auto-observer"
      />
    </div>
    <div
      class="simplebar-mask"
    >
      <div
        class="simplebar-offset"
      >
        <div
          class="simplebar-content-wrapper"
        >
          <div
            class="simplebar-content"
          >
            
            <div
              class="test"
            >
              Infinite scroll content...
            </div>
            
            <div
              class="ik-scroll-area__loader"
            >
              <ik-loader-circle-stub
                design="default"
                indeterminate="true"
                rotate="0"
                size="20"
                thickness="2"
                value="0"
              />
            </div>
            <!--v-if-->
          </div>
        </div>
      </div>
    </div>
    <div
      class="simplebar-placeholder"
    />
  </div>
  <div
    class="simplebar-track simplebar-horizontal"
  >
    <div
      class="simplebar-scrollbar"
    />
  </div>
  <div
    class="simplebar-track simplebar-vertical"
  >
    <div
      class="simplebar-scrollbar"
    />
  </div>
</div>
`;

exports[`IkScrollArea renders loading in reverse mode 1`] = `
<div
  class="ik-scroll-area"
  data-simplebar=""
  data-simplebar-auto-hide="false"
>
  <div
    class="simplebar-wrapper"
  >
    <!--v-if-->
    <div
      class="simplebar-height-auto-observer-wrapper"
    >
      <div
        class="simplebar-height-auto-observer"
      />
    </div>
    <div
      class="simplebar-mask"
    >
      <div
        class="simplebar-offset"
      >
        <div
          class="simplebar-content-wrapper"
        >
          <div
            class="simplebar-content"
          >
            <!--v-if-->
            <div
              class="ik-scroll-area__loader"
            >
              <ik-loader-circle-stub
                design="default"
                indeterminate="true"
                rotate="0"
                size="20"
                thickness="2"
                value="0"
              />
            </div>
            
            <div
              class="test"
            >
              Infinite scroll content...
            </div>
            
          </div>
        </div>
      </div>
    </div>
    <div
      class="simplebar-placeholder"
    />
  </div>
  <div
    class="simplebar-track simplebar-horizontal"
  >
    <div
      class="simplebar-scrollbar"
    />
  </div>
  <div
    class="simplebar-track simplebar-vertical"
  >
    <div
      class="simplebar-scrollbar"
    />
  </div>
</div>
`;

exports[`IkScrollArea renders with correct default state 1`] = `
<div
  class="ik-scroll-area"
  data-simplebar=""
  data-simplebar-auto-hide="false"
>
  <div
    class="simplebar-wrapper"
  >
    <!--v-if-->
    <div
      class="simplebar-height-auto-observer-wrapper"
    >
      <div
        class="simplebar-height-auto-observer"
      />
    </div>
    <div
      class="simplebar-mask"
    >
      <div
        class="simplebar-offset"
      >
        <div
          class="simplebar-content-wrapper"
        >
          <div
            class="simplebar-content"
          >
            
            
            <!--v-if-->
            <!--v-if-->
          </div>
        </div>
      </div>
    </div>
    <div
      class="simplebar-placeholder"
    />
  </div>
  <div
    class="simplebar-track simplebar-horizontal"
  >
    <div
      class="simplebar-scrollbar"
    />
  </div>
  <div
    class="simplebar-track simplebar-vertical"
  >
    <div
      class="simplebar-scrollbar"
    />
  </div>
</div>
`;

exports[`IkScrollArea renders with default slot 1`] = `
<div
  class="ik-scroll-area"
  data-simplebar=""
  data-simplebar-auto-hide="false"
>
  <div
    class="simplebar-wrapper"
  >
    <!--v-if-->
    <div
      class="simplebar-height-auto-observer-wrapper"
    >
      <div
        class="simplebar-height-auto-observer"
      />
    </div>
    <div
      class="simplebar-mask"
    >
      <div
        class="simplebar-offset"
      >
        <div
          class="simplebar-content-wrapper"
        >
          <div
            class="simplebar-content"
          >
            
            <div
              class="test"
            >
              Area scroll content...
            </div>
            
            <!--v-if-->
            <!--v-if-->
          </div>
        </div>
      </div>
    </div>
    <div
      class="simplebar-placeholder"
    />
  </div>
  <div
    class="simplebar-track simplebar-horizontal"
  >
    <div
      class="simplebar-scrollbar"
    />
  </div>
  <div
    class="simplebar-track simplebar-vertical"
  >
    <div
      class="simplebar-scrollbar"
    />
  </div>
</div>
`;
